python/ray/data/_internal/batcher.py

```diff
@@ -275,46 +267,63 @@ def has_batch(self) -> bool:
         buffer_size = self._buffer_size()
 
         if not self._done_adding:
-            # If still adding blocks, ensure that removing a batch wouldn't cause the
-            # shuffle buffer to dip beneath its configured minimum size.
-            return buffer_size - self._batch_size >= self._buffer_min_size
+            # Delay pulling of batches until the buffer is large enough in order to
+            # amortize compaction overhead.
+            return (
+                buffer_size - self._batch_size
+                >= self._buffer_min_size * SHUFFLE_BUFFER_COMPACTION_RATIO
+            )
         else:
             return buffer_size >= self._batch_size
 
     def _buffer_size(self) -> int:
         """Return shuffle buffer size."""
         buffer_size = self._builder.num_rows()
-        if self._shuffle_buffer is not None:
-            # Include the size of the concrete (materialized) shuffle buffer, adjusting
-            # for the batch head position, which also serves as a counter of the number
-            # of already-yielded rows from the current concrete shuffle buffer.
-            buffer_size += (
-                BlockAccessor.for_block(self._shuffle_buffer).num_rows()
-                - self._batch_head
-            )
+        buffer_size += self._materialized_buffer_size()
         return buffer_size
 
+    def _materialized_buffer_size(self) -> int:
+        """Return materialized (compacted portion of) shuffle buffer size."""
+        if self._shuffle_buffer is None:
+            return 0
+        # The size of the concrete (materialized) shuffle buffer, adjusting
+        # for the batch head position, which also serves as a counter of the number
+        # of already-yielded rows from the current concrete shuffle buffer.
+        return max(
+            0,
+            BlockAccessor.for_block(self._shuffle_buffer).num_rows() - self._batch_head,
+        )
+
     def next_batch(self) -> Block:
         """Get the next shuffled batch from the shuffle buffer.
 
         Returns:
             A batch represented as a Block.
         """
         assert self.has_batch() or (self._done_adding and self.has_any())
-        # Add rows in the builder to the shuffle buffer.
-        if self._builder.num_rows() > 0:
+        # Add rows in the builder to the shuffle buffer. Note that we delay compaction
+        # as much as possible to amortize the concatenation overhead. Compaction is
+        # only necessary when the materialized buffer size falls below the min size.
+        if self._builder.num_rows() > 0 and (
+            self._done_adding
+            or self._materialized_buffer_size() <= self._buffer_min_size
```

Comment:
I think this should be `self._materialized_buffer_size() <= self._batch_size`.
Because the data in the materialized buffer `self._shuffle_buffer` is already shuffled. We can directly take a batch from there as long as there is enough data. 
This change almost doubled the throughput on my laptop for the benchmark in the PR description. 

Also, after this change, `SHUFFLE_BUFFER_COMPACTION_RATIO` doesn't have significant impacts in my local experiments.

Reply:
I think this is important for shuffle quality though. It's not enough to materialize a batch of data, we need to ensure the shuffling spans at least the min buffer size.

Reply:
my understanding like this. we keep adding new data to `self._builder`. when its size reaches `_buffer_min_size`, we generate a new `_shuffle_buffer` from the builder. Thus, it means that the data in `self._shuffle_buffer` is already shuffled, spanning the min buffer size. After that, we can keep taking batches from `_shuffle_buffer` until the remaining data is less than a batch.

Reply:
The issue is that that is not enough to implement a true sliding shuffle window of that size. For example, suppose you have 10 records, and a shuffle window size of 5, and a batch size of 2.

input = [1, 2, 3, 4, 5, 6, 7, 8, 9 10].

With the naive approach, you might shuffle only [1, 2, 3, 4, 5] and [6, 7, 8, 9, 10] separately, but this isn't random enough. You actually want to shuffle something like [1, 2, 3, 4, 5], and then [3, 4, 5, 6, 7], and then [5, 6, 7, 8, 9], i.e., sliding window.

Reply:
I see. The code makes sense to me now. thanks. 

2 small remaining issues: 

1. the logic in `has_batch` can be optimized as the following. Because when `self._materialized_size() >= self._min_buffer_size `, we can take a batch directly from the buffer without trigger compaction. this should only improve latency, not throughput. 

```
           self._materialized_size() >= self._min_buffer_size  or (
                buffer_size - self._batch_size
                >= self._buffer_min_size * SHUFFLE_BUFFER_COMPACTION_RATIO
            )
 ```
 
 2. With compact ratio = 1.5, if we change the batch_size to 800 for the benchmark. The code will crash due to the `can_add` issue I mentioned before. better add a test with different combinations of batch_size + min_buffer_size.